samples/event-sourcing-esdb-simple/src/main/java/io/eventdriven/ecommerce/shoppingcarts/removingproductitem/RemoveProductItemFromShoppingCart.java

```diff
@@ -0,0 +1,41 @@
+package io.eventdriven.ecommerce.shoppingcarts.removingproductitem;
+
+import io.eventdriven.ecommerce.shoppingcarts.Events;
+import io.eventdriven.ecommerce.shoppingcarts.ShoppingCart;
+import io.eventdriven.ecommerce.shoppingcarts.productitems.PricedProductItem;
+
+import java.util.UUID;
+
+public record RemoveProductItemFromShoppingCart(
+  UUID shoppingCartId,
+  PricedProductItem productItem,
+  Long expectedVersion
+) {
+  public static RemoveProductItemFromShoppingCart of(UUID cartId, PricedProductItem productItem, Long expectedVersion) {
+    if (cartId == null)
+      throw new IllegalArgumentException("Cart id has to be defined");
+
+    if (productItem == null)
+      throw new IllegalArgumentException("Product item has to be defined");
+
+    if (expectedVersion == null)
+      throw new IllegalArgumentException("Expected version has to be provided");
+
+    return new RemoveProductItemFromShoppingCart(cartId, productItem, expectedVersion);
+  }
+
+  public static Events.ProductItemRemovedFromShoppingCart handle(
+    RemoveProductItemFromShoppingCart command,
+    ShoppingCart shoppingCart
+  ) {
+    if (shoppingCart.isClosed())
+      throw new IllegalStateException("Adding product item for cart in '%s' status is not allowed.".formatted(shoppingCart.status()));
+
+    shoppingCart.productItems().remove(command.productItem());
```

Comment:
It's very strange for me that you're modifying the state here. It's a side effect, but I'd rather go in the direction of `pure functional command handlers`. Do we need this modification with with Event Sourcing? I think not. Event is enough. If you really want a new state you can return from the method events + newState (immutable object). Then you're method will not have side effects.

Reply:
In fact, if you have a look at the ProductList class, it's immutable, but I think that I'll do this more explicit and simplified 👍

Reply:
OK, "assertThatCanRemove" makes sens! :) 
I thought that there was a code, where you modify state and use the side effect.
Then you said it's immutable which was totally mindfuck, because "why he return some value and do nothing with tat".

Now I'm only not sure if we need return value from `assertThatCanRemove`. Is it usable somewhere?

Reply:
Yes, it's usable internally in the remove function, that's used in state rehydration.